lib/pos_blockchain.go

```diff
@@ -0,0 +1,1134 @@
+			return RuleErrorPoSTimeoutBlockViewNotGreaterThanParent
+		}
+	}
+	return nil
+}
+
+// hasValidBlockProposerPoS validates that the proposer is the expected proposer for the
+// block height + view number pair. It returns a bool indicating whether
+// we confirmed that the leader is valid. If we receive an error, we are unsure
+// if the leader is invalid or not, so we return false.
+func (bc *Blockchain) hasValidBlockProposerPoS(block *MsgDeSoBlock) (_isValidBlockProposer bool, _err error) {
+	utxoView, err := bc.getUtxoViewAtBlockHash(*block.Header.PrevBlockHash)
+	if err != nil {
+		return false, errors.Wrapf(err, "hasValidBlockProposerPoS: Problem initializing UtxoView")
+	}
+	currentEpochEntry, err := utxoView.GetCurrentEpochEntry()
+	if err != nil {
+		return false, errors.Wrapf(err, "hasValidBlockProposerPoS: Problem getting current epoch entry")
+	}
+	leaders, err := utxoView.GetSnapshotLeaderSchedule()
+	if err != nil {
+		return false, errors.Wrapf(err, "hasValidBlockProposerPoS: Problem getting leader schedule")
+	}
+	if len(leaders) == 0 {
+		return false, errors.Wrapf(err, "hasValidBlockProposerPoS: No leaders found in leader schedule")
+	}
+	if block.Header.Height < currentEpochEntry.InitialBlockHeight {
+		return false, nil
+	}
+	if block.Header.ProposedInView < currentEpochEntry.InitialView {
+		return false, nil
+	}
+	heightDiff := block.Header.Height - currentEpochEntry.InitialBlockHeight
+	viewDiff := block.Header.ProposedInView - currentEpochEntry.InitialView
+	if viewDiff < heightDiff {
+		return false, nil
+	}
+	// We compute the current index in the leader schedule as follows:
+	// [(block.View - currentEpoch.InitialView) - (block.Height - currentEpoch.InitialHeight)] % len(leaders)
+	// The number of views that have elapsed since the start of the epoch is block.View - currentEpoch.InitialView.
+	// The number of blocks that have been added to the chain since the start of the epoch is block.Height - currentEpoch.InitialHeight.
+	// The difference between these two numbers is the number of timeouts that have occurred in this epoch.
+	// For each timeout, we need to go to the next leader in the schedule.
+	// If we have more timeouts than leaders in the schedule, we start from the top of the schedule again,
+	// which is why we take the modulo of the length of the leader schedule.
+	// A quick example: If we have 3 leaders in the schedule and the epoch started at height 10 and view 11,
+	// and the current block is at height 15 and view 17, then the number of timeouts that have occurred is
+	// (17 - 11) - (15 - 10) = 1. This means this block should be proposed by the 2nd leader in the schedule,
+	// which is at index 1.
+	leaderIdxUint64 := (viewDiff - heightDiff) % uint64(len(leaders))
+	if leaderIdxUint64 > math.MaxUint16 {
+		return false, nil
+	}
+	leaderIdx := uint16(leaderIdxUint64)
+	leaderEntry, err := utxoView.GetSnapshotLeaderScheduleValidator(leaderIdx)
```

Comment:
So I notice the utxoView is set for the PrevBlock, but it looks like you're trying to use it to get the leader for the *current* block. I think that works as long as the function is returning the leader schedule for the *next* block, which I think it is. Otherwise, this could get messed up on epoch boundaries where the leader schedule is changing. Just double-check it.

Reply:
When we connect the parent block, if it's the last block in the epoch, the end of epoch hook runs and the CurrentEpochEntry is updated so we have the correct leader schedule & validator set for the next block.

Reply:
Awesome sounds good. Yeah I figured this function is telling you what the *next* leader is going to be.